File: simple-stack-example-mvp/src/main/java/com/zhuinden/simplestackdemoexamplemvp/presentation/paths/statistics/StatisticsPresenter.java
package com.zhuinden.simplestackdemoexamplemvp.presentation.paths.statistics;

import com.zhuinden.simplestackdemoexamplemvp.data.repository.TaskRepository;
import com.zhuinden.simplestackdemoexamplemvp.presentation.objects.Task;
import com.zhuinden.simplestackdemoexamplemvp.util.BasePresenter;

import org.javatuples.Pair;

import java.util.List;

import javax.inject.Inject;

import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by Owner on 2017. 01. 27..
 */

public class StatisticsPresenter
        extends BasePresenter<StatisticsView, StatisticsPresenter> {
    @Inject
    public StatisticsPresenter() {
    }

    @Inject
    TaskRepository tasksRepository;

    Subscription subscription;

    @Override
    protected void onAttach(StatisticsView view) {
        subscription = Observable.combineLatest(tasksRepository.getActiveTasks(), //
                tasksRepository.getCompletedTasks(), //
                (activeTasks, completedTasks) -> Pair.with(activeTasks, completedTasks)) //
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(pairOfActiveAndCompletedTasks -> {
                    List<Task> activeTasks = pairOfActiveAndCompletedTasks.getValue0();
                    List<Task> completedTasks = pairOfActiveAndCompletedTasks.getValue1();
                    if(getView() != null) {
                        getView().showStatistics(activeTasks.size(), completedTasks.size());
                    }
                });
    }

    @Override
    protected void onDetach(StatisticsView view) {
        subscription.unsubscribe();
    }
}

File: samples/legacy-architecture-samples/simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/features/statistics/StatisticsPresenter.kt
package com.zhuinden.simplestackdemoexamplemvp.features.statistics

import com.zhuinden.rxcombinetuplekt.combineTuple
import com.zhuinden.simplestackdemoexamplemvp.core.mvp.BasePresenter
import com.zhuinden.simplestackdemoexamplemvp.data.repository.TaskRepository
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

/**
 * Created by Owner on 2017. 01. 27..
 */

class StatisticsPresenter(
    private val tasksRepository: TaskRepository
) : BasePresenter<StatisticsView>(), StatisticsView.Presenter {
    lateinit var disposable: Disposable

    override fun onAttach(view: StatisticsView) {
        disposable = combineTuple(
            tasksRepository.activeTasksWithChanges,
            tasksRepository.completedTasksWithChanges
        ).subscribeOn(Schedulers.computation())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { (activeTasks, completedTasks) ->
                view.showStatistics(activeTasks.size, completedTasks.size)
            }
    }

    override fun onDetach(view: StatisticsView) {
        disposable.dispose()
    }
}
